### ISSUE TITLE
MueLu doesn't compile....

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#3378]
MueLu RefMaxwell doesn't compile 

@trilinos/muelu 

@cgcgcg 

## Current Behavior

```
[0/97] Building CXX object packages/muelu/src/CMakeFiles/muelu.dir/Utils/ExplicitInstantiation/MueLu_CoalesceDropFactory_kokkos.cpp.o
p
In file included from packages/muelu/src/MueLu_Utilities_kokkos.hpp:1:0,
                 from /home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:63,
                 from packages/muelu/src/Utils/ExplicitInstantiation/MueLu_CoalesceDropFactory_kokkos.cpp:55:
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_Utilities_kokkos_decl.hpp:276:63: error: ‘RealValuedMultiVector’ was not declared in this scope
     static RCP<MultiVector> RealValuedToScalarMultiVector(RCP<RealValuedMultiVector> X);
                                                               ^~~~~~~~~~~~~~~~~~~~~
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_Utilities_kokkos_decl.hpp:276:84: error: template argument 1 is invalid
     static RCP<MultiVector> RealValuedToScalarMultiVector(RCP<RealValuedMultiVector> X);
                                                                                    ^
In file included from packages/muelu/src/MueLu_Utilities_kokkos.hpp:1:0,
                 from /home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:63,
                 from packages/muelu/src/Utils/ExplicitInstantiation/MueLu_CoalesceDropFactory_kokkos.cpp:55:
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_Utilities_kokkos_decl.hpp:415:63: error: ‘RealValuedMultiVector’ was not declared in this scope
     static RCP<MultiVector> RealValuedToScalarMultiVector(RCP<RealValuedMultiVector> X);
                                                               ^~~~~~~~~~~~~~~~~~~~~
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_Utilities_kokkos_decl.hpp:415:84: error: template argument 1 is invalid
     static RCP<MultiVector> RealValuedToScalarMultiVector(RCP<RealValuedMultiVector> X);
                                                                                    ^
In file included from /home/mbetten/Trilinos/Trilinos/packages/kokkos/core/src/Kokkos_Core_fwd.hpp:51:0,
                 from /home/mbetten/Trilinos/Trilinos/packages/kokkos/core/src/Kokkos_Core.hpp:50,
                 from /home/mbetten/Trilinos/Trilinos/packages/teuchos/kokkoscompat/src/KokkosCompat_ClassicNodeAPI_Wrapper.hpp:5,
                 from /home/mbetten/Trilinos/Trilinos/packages/tpetra/classic/NodeAPI/Kokkos_DefaultNode.hpp:47,
                 from /home/mbetten/Trilinos/Trilinos/packages/muelu/src/Headers/MueLu_ConfigDefs.hpp:54,
                 from packages/muelu/src/Utils/ExplicitInstantiation/MueLu_CoalesceDropFactory_kokkos.cpp:49:
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp: In instantiation of ‘struct MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>’:
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:643:33:   required from ‘void MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]’
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_ETI_4arg.hpp:42:3:   required from here
/home/mbetten/Trilinos/Trilinos/packages/kokkos/core/src/Kokkos_Macros.hpp:213:27: warning: ‘MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>’ has a field ‘MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>::<distFunctor capture>’ whose type uses the anonymous namespace [-Wsubobject-linkage]
   #define KOKKOS_LAMBDA [=]
                           ^
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:644:15: note: in expansion of macro ‘KOKKOS_LAMBDA’
               KOKKOS_LAMBDA(const LO row) {
               ^
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp: In instantiation of ‘struct MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = long long int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>’:
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:643:33:   required from ‘void MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = long long int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]’
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Utils/MueLu_ETI_4arg.hpp:42:3:   required from here
/home/mbetten/Trilinos/Trilinos/packages/kokkos/core/src/Kokkos_Macros.hpp:213:27: warning: ‘MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = long long int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>’ has a field ‘MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Build(MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level&) const [with Scalar = double; LocalOrdinal = int; GlobalOrdinal = long long int; DeviceType = Kokkos::OpenMP; typename DeviceType::memory_space = Kokkos::HostSpace; MueLu::CoalesceDropFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Kokkos::Compat::KokkosDeviceWrapperNode<DeviceType> >::Level = MueLu::Level]::<lambda(MueLu::CoalesceDropFactory_kokkos<double, int, long long int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::OpenMP> >::LO)>::<distFunctor capture>’ whose type uses the anonymous namespace [-Wsubobject-linkage]
   #define KOKKOS_LAMBDA [=]
                           ^
/home/mbetten/Trilinos/Trilinos/packages/muelu/src/Graph/MueLu_CoalesceDropFactory_kokkos_def.hpp:644:15: note: in expansion of macro ‘KOKKOS_LAMBDA’
               KOKKOS_LAMBDA(const LO row) {
               ^
ninja: build stopped: subcommand failed.
dir-opt (gcc) 212 $ 


```

## Your Environment

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by bathmatt** [comment]
Whoops, maybe it wasn't @cgcgcg who broke it.  a334f7fade1588d3bb617c41eedd75284d3ec65c is the bad commit..

**Comment by lucbv** [comment]
@bathmatt I did this and know where the fix needs to go so I'll take care of it.
It should be done by the end of the day.

**Comment by lucbv** [comment]
@bathmatt let me know if PR #3379 fixed the pb.

**Comment by ikalash** [comment]
This issue is also happening on the NERSC machines (Cori, Edison) in the Albany dashboard starting Sat. Sept. 1: https://my.cdash.org/viewBuildError.php?buildid=1515008 .

**Comment by lucbv** [comment]
@ikalash I also see an issue with some complex tests, I am looking into this, it is rather easy to fix, the main problem is finding all the places suffering from this issue. I will post again when I get the new PR fixing the problem with complex merged in.

**Comment by ikalash** [comment]
Thanks @lucbv.  I'll keep an eye out for your PR.  It is sort of curious that this issue shows up in the NERSC Albany nightlies but it seems not the others, e.g., those on my workstation.  I did not compare the configure scripts - they are not identical, nor are the compilers; perhaps there is something in that.

**Comment by lucbv** [comment]
@bathmatt, let us wait to get a green light from @ikalash too since she also observed issues in her builds. I have a PR merged yesterday for complex builds and I have one more this morning for a strange ETI case. This last one should completely clear the MueLu dashboard so hopefully that's the end of it.

**Comment by bathmatt** [comment]
@ikalash you can close this when you've confirmed things, 
Thanks
Matt

**Comment by ikalash** [comment]
@lucbv, @bathmatt : I am working on this right now.  Last night the Cori and Edison tests failed again.  I am pulling Trilinos now and re-trying.

**Comment by ikalash** [comment]
I haven't been following this very carefully so I think I missed something: is the fix in a PR or in develop already?  It is in @lucbv 's branch?

**Comment by lucbv** [comment]
@ikalash a fix was merged on Sunday evening and I am adding a final round today, depending on how you build Trilinos you might need to wait until PR #3384 is merged in.
If the auto-tester works today it should not take too long.

**Comment by ikalash** [comment]
@lucbv: OK this is what I thought originally but I wanted to double check.  In this case, the issue is not fixed on NERSC  unfortunately.  Here is a snippet of the error: 
```
[ 96%] Building CXX object packages/muelu/src/CMakeFiles/muelu.dir/Utils/ExplicitInstantiation/MueLu_AggregationPhase3Algorithm.cpp.o
In file included from /project/projectdirs/piscees/nightlyEdisonCDash/repos/Trilinos/packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_def.hpp:52:0,
                 from /global/project/projectdirs/piscees/nightlyEdisonCDash/build/TriBuild/packages/muelu/src/Utils/ExplicitInstantiation/MueLu_Zoltan2Interface.cpp:55:
/project/projectdirs/piscees/nightlyEdisonCDash/repos/Trilinos/packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_decl.hpp: In member function 'virtual void MueLu::Zoltan2Interface<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >::Build(MueLu::Zoltan2Interface<double, int, int, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial> >::Level&) const':
/project/projectdirs/piscees/nightlyEdisonCDash/repos/Trilinos/packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_decl.hpp:210:51: error: 'RealValuedMultiVector' was not declared in this scope
       level_->Set("Coordinates",          Get<RCP<RealValuedMultiVector> >(currentLevel, "Coordinates"));
                                                   ^~~~~~~~~~~~~~~~~~~~~
/project/projectdirs/piscees/nightlyEdisonCDash/repos/Trilinos/packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_decl.hpp:210:51: note: suggested alternative: 'BlockedMultiVector'
       level_->Set("Coordinates",          Get<RCP<RealValuedMultiVector> >(currentLevel, "Coordinates"));
                                                   ^~~~~~~~~~~~~~~~~~~~~
                                                   BlockedMultiVector
/project/projectdirs/piscees/nightlyEdisonCDash/repos/Trilinos/packages/muelu/src/Rebalancing/MueLu_Zoltan2Interface_decl.hpp:210:72: error: template argument 1 is invalid
       level_->Set("Coordinates",          Get<RCP<RealValuedMultiVector> >(currentLevel, "Coordinates"));
```
The issue also is showing up in the Albany Ride build, which I just reactivated today: http://cdash.sandia.gov/CDash-2-3-0/viewBuildError.php?buildid=75589 .  

I thought when I first saw the issue that maybe I was supposed to checkout a separate branch to be turned into a PR, hence the confusion.  Please let me know when the PR you're mentioning is merged in, and I can try the tests again.

**Comment by lucbv** [comment]
@ikalash that's good, the current PR is fixing this particular issue so whenever the auto-tester tests it the issue will be resolved. I will let you know when the merge happens.
Thanks!

**Comment by lucbv** [comment]
@ikalash the new PR has just been merged, let me know when you have time to check if it fixes the issue properly for your build.
Thanks!

**Comment by lucbv** [comment]
Yeah, no problem we can look at it tomorrow, there is also a MueLu nightly reporting this build issue, I will monitor it tomorrow morning too and hopefully we can close this.

**Comment by ikalash** [comment]
Sounds like a plan!

**Comment by ikalash** [comment]
Albany night lies are back to normal.  Thanks for fixing this!
